### PR TITLE
nspawn: do not fail if udev is not running

### LINKED ISSUE #14223: nspawn: --network-interface fails with "Network interface is not initialized yet." on network interfaces created with iproute2
**systemd version the issue has been seen with**
> systemd-v244-rc1

**Used distribution**
 > Ubuntu Focal

**Expected behaviour you didn't see**
 > systemd-nspawn starts the container and inherits the veth interface created with iproute2 

**Unexpected behaviour you saw**
 > systemd-nspawn fails with "Network interface is not initialized yet."

**Steps to reproduce the problem**

```
ip link add name vi1 type veth peer name vi2
systemd-nspawn -b -D <rootfs> --network-interface vi1
```

### PR DESCRIPTION
**Pull request by yuwata**
If nspawn is running container, usually udev is not running. In such a
case, let's assume that the interface is already initialized.
Also, this makes nspawn refuse to use the network interface which is
under renaming.

Fixes #14223.

### CODE REVIEW

File: src/nspawn/nspawn-network.c [hunk 1]
```diff
@@ -395,24 +397,33 @@ int remove_bridge(const char *bridge_name) {
...
+
+                r = sd_device_get_is_initialized(d);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to determine whether interface %s is initialized or not: %m", name);
+                if (r == 0)
+                        return log_error_errno(SYNTHETIC_ERRNO(EBUSY), "Network interface %s is not initialized yet.", name);
+
+                r = device_is_renaming(d);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to determine the interface %s is renamed or not: %m", name);
```
Comment (keszybz) [inline src/nspawn/nspawn-network.c]: Maybe: "Failed to determine if interface is *being* renamed: %m".

File: src/nspawn/nspawn-network.c [hunk 2]
```diff
@@ -395,24 +397,33 @@ int remove_bridge(const char *bridge_name) {
...
+                /* not in a container. udev will be around. */
+
+                sprintf(ifi_str, "n%i", ifi);
+                r = sd_device_new_from_device_id(&d, ifi_str);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to get device for interface %s: %m", name);
+
+                r = sd_device_get_is_initialized(d);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to determine whether interface %s is initialized or not: %m", name);
```
Comment (keszybz) [inline src/nspawn/nspawn-network.c]: Drop "or not", the meaning is the same without that.

File: src/nspawn/nspawn-network.c [hunk 3]
```diff
@@ -395,24 +397,33 @@ int remove_bridge(const char *bridge_name) {
...
+                if (r == 0)
+                        return log_error_errno(SYNTHETIC_ERRNO(EBUSY), "Network interface %s is not initialized yet.", name);
+
+                r = device_is_renaming(d);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to determine the interface %s is renamed or not: %m", name);
+                if (r > 0)
+                        return log_error_errno(SYNTHETIC_ERRNO(EBUSY), "Network Interface %s is under renaming.", name);
```
Comment (keszybz) [inline src/nspawn/nspawn-network.c]: "Interface %s is being renamed."

File: src/nspawn/nspawn-network.c [hunk 4]
```diff
@@ -395,24 +397,33 @@ int remove_bridge(const char *bridge_name) {
...
-                return log_error_errno(r, "Failed to determine whether interface %s is initialized or not: %m", name);
-        if (r == 0) {
-                log_error("Network interface %s is not initialized yet.", name);
-                return -EBUSY;
+                /* not in a container. udev will be around. */
+
+                sprintf(ifi_str, "n%i", ifi);
+                r = sd_device_new_from_device_id(&d, ifi_str);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to get device for interface %s: %m", name);
```
Comment (keszybz) [inline src/nspawn/nspawn-network.c]: Better I think because more precise: "Failed to get device %s: %m", ifi

**Comment by poettering** [comment]
hmm, so udevd is conditioned like this: "ConditionPathIsReadWrite=/sys", i.e. the idea was that container managers or any other env where device mgmt by udev is not desired would mount /sys read-only, and thus get no udev. I think we should use the same, matching condition check, i.e. don#t wait for udev if path_is_read_only_fs() > 0.

**Comment by yuwata** [comment]
@poettering and @keszybz Thank you for the comments. I've force-pushed the revised version. I've also updated networkd, as the code introduced to nspawn is copied from networkd.

File: src/network/networkd-link.c [hunk 5]
```diff
@@ -3314,11 +3314,11 @@ int link_add(Manager *m, sd_netlink_message *message, Link **ret) {
 
                 r = device_is_renaming(device);
                 if (r < 0) {
-                        log_link_warning_errno(link, r, "Failed to determine the device is renamed or not: %m");
+                        log_link_warning_errno(link, r, "Failed to determine the device is being renamed: %m");
                         goto failed;
                 }
                 if (r > 0) {
-                        log_link_debug(link, "Interface is under renaming, pending initialization.");
+                        log_link_debug(link, "Interface is under being renamed, pending initialization.");
```
Comment (dnicolodi) [inline src/network/networkd-link.c]: "Interface is being renamed"

**Comment by DaanDeMeyer** [comment]
Tested the pr locally by starting a nested container from within another container that did not have udev running with `--network-interface=<intf-by-iproute2>` and can confirm it fixes the issue.

File: src/network/networkd-link.c (hunk 5 above)
Comment (yuwata) [reply src/network/networkd-link.c]: Thanks. Fixed.

**Comment by yuwata** [comment]
@DaanDeMeyer Thank you for testing this PR.

**Pull request merged**